README.md
<img src="https://avatars.githubusercontent.com/u/83244875?s=400&v=4" width="50"/>

Business Logic - Calculator
=====================

This business logic is an example how to implement a custom logic for your chatbot on the [Business Bot Platform](https://www.citunius.de/). This project implements a simple calculator and allows developers to use it as template for their own business logic and run it on the Business Bot platform.

A business logic is a service that is connected to a chatbot in order to provide a specific service for mobile users. The chatbot defines how the message is transmitted, while the business logic decides how the message is processed.

The business logic is effectively the WebApp while the Business Bot Platform is the application server such as Tomcat.


Features/Benefits
-------------------
- Processing basic arithmetic tasks
- Allows whitespaces in a task

Please check out the [documentation](https://library.citunius.de/products/bbp_edition_community/) for
additional information on this project.

JDK 1.8 is required. Make sure your default JDK version is >=1.8
by typing `java -version`.


## How to use


## Compile
Use the [Business Logic - Creator](https://github.com/citunius/businesslogic.creator) to compile this project. Once compiled, deploy the business logic package on the Business Bot Platform as documented [here](https://library.citunius.de/products/bbp_edition_community/documentation/R2021-FP2116/en/administration/botadministration/businesslogics/#add-business-logic-from-a-local-file).


## Business Bot Platform Documentation
For a more detailed documentation about the Business Bot Platform, please visit: [https://library.citunius.de/products/bbp_edition_community/](https://library.citunius.de/products/bbp_edition_community/) 


# History
Initial version by [Citunius GmbH](https://www.citunius.de/)

## License
This software is under [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0).

## Suggestions
Any suggestions are  welcome.

CHANGELOG.md
## Change Log
All notable changes to this project will be documented in this file.

### [Unreleased][unreleased]

#### Fixed

#### Changed

#### Added

#### Removed

### [1.0.0] - 2021-05-03

Initial Version.



src/test/de/citunius/businesslogic/Main.java
package test.de.citunius.businesslogic;

import java.util.Date;
import java.util.HashMap;

import de.citunius.bbp.im.gm.objects.Data;
import de.citunius.bbp.im.gm.objects.Message;
import de.citunius.bbp.objects.AnonymousUserAccount;
import de.citunius.bbp.objects.MobileAppProvider;
import de.citunius.businesslogic.calculator.CalculatorBot;
import de.citunius.businesslogic.calculator.plugin.PluginClient.BusinessLogic;
import de.citunius.businesslogicapi.common.ConstantsPlugin;

public class Main {

	// API account credentials (if business logic want to call the BBP webservices)
	public static String BBP_APIACCOUNT_APIID_PARAM = "BusinessBotPlatform.Account.ApiId";
	public static String BBP_APIACCOUNT_APIKEY_PARAM = "BusinessBotPlatform.Account.ApiKey";
	
	// Test Values
	public static String TENANTID = "demo";
	public static String ACCOUNTID = "ALL_ACCOUNTS";
	
	public static String USERNAME = "john";
	public static String FIRSTNAME = "John";
	public static String LASTNAME = "Doe";
	
	public static String MESSENGER_NAME = "Telegram";
	public static String PLUGIN_ID = "1";
	public static String PLUGIN_NAME = "demo.ALL_ACCOUNTS.de.citunius.generic.calculator";
	public static String PLUGIN_WEBUI_LANGUAGE = "en";
	public static String PLUGIN_RESOURCE_PATH = "BusinessLogic.Component.ConversationalEditor\\resources";
	public static String BBP_WEBSERVICES_URL = "https://demo.bbp.local:450/bbp/rest/api/1.0/";
	public static String PLUGIN_WORKDIR = "C:\\temp";
	public static String PLUGIN_SYSTEMDIR = "C:\\Citunius.DevEnv\\eclipse-workspace-java";
	public static String INSTANTMESSENGER_FILESTORE_PATH = "c:\\temp\\bbp\\filestore";
	public static String BBP_APIACCOUNT_APIID = "u-98edf8d";
	public static String BBP_APIACCOUNT_APIKEY = "5407807";
			
	public static void main(String[] args) {
		String userMessage = "4 + 2";
		
		BusinessLogic bl = new BusinessLogic();
		HashMap<String, String> pluginMap = getPluginMap();
		
		de.citunius.bbp.im.gm.objects.User fromUser = new de.citunius.bbp.im.gm.objects.User();
		fromUser.setUserName("john");
		fromUser.setFirstName("John");
		fromUser.setLastName("Doe");
		
		de.citunius.bbp.im.gm.objects.User toUser = new de.citunius.bbp.im.gm.objects.User();
		toUser.setUserName("jane");
		toUser.setFirstName("jane");
		toUser.setLastName("Doe");
		Data data = new Data();
		
		Message message = new Message();
		message.setFrom(fromUser);
		message.setTo(toUser);
		message.setData(data);				
		message.setText(userMessage);

		boolean hasAnonymousUserAccount = true;
		MobileAppProvider mobileAppProvider = new MobileAppProvider(TENANTID, ACCOUNTID, 0, MESSENGER_NAME, MESSENGER_NAME, "Enabled");
		AnonymousUserAccount anonymousUserAccount = new AnonymousUserAccount(TENANTID, ACCOUNTID, 0, USERNAME, FIRSTNAME, LASTNAME, mobileAppProvider, new Date());
		boolean hasMobileUserAccount = false;
		String jsonMobileUserAccount = null;
		
		System.out.println("Sending user message: "+userMessage);
		String result = bl.handleIncomingMessage(TENANTID, ACCOUNTID, pluginMap, message.toJson().toString(), hasAnonymousUserAccount, anonymousUserAccount.toJson().toString(), hasMobileUserAccount, jsonMobileUserAccount);
		System.out.println("Business Logic replied: "+result);
	}
	
	/**
	 * Compose plugin map passed to business logic
	 * 
	 * @return
	 */
	public static HashMap<String, String> getPluginMap() {
		HashMap<String, String> pluginMap = new HashMap<>();
		pluginMap.put(ConstantsPlugin.MESSENGER_NAME, MESSENGER_NAME);
		pluginMap.put(ConstantsPlugin.PLUGIN_ID, PLUGIN_ID);
		pluginMap.put(ConstantsPlugin.PLUGIN_NAME, PLUGIN_NAME);
		
		pluginMap.put(ConstantsPlugin.TENANTID, TENANTID);
		pluginMap.put(ConstantsPlugin.ACCOUNTID, ACCOUNTID);
		pluginMap.put(ConstantsPlugin.PLUGIN_WEBUI_LANGUAGE, PLUGIN_WEBUI_LANGUAGE);
		pluginMap.put(ConstantsPlugin.PLUGIN_RESOURCE_PATH, PLUGIN_RESOURCE_PATH);
		pluginMap.put(ConstantsPlugin.BBP_WEBSERVICES_URL, BBP_WEBSERVICES_URL);
		pluginMap.put(ConstantsPlugin.PLUGIN_WORKDIR, PLUGIN_WORKDIR);
		pluginMap.put(ConstantsPlugin.PLUGIN_SYSTEMDIR, PLUGIN_SYSTEMDIR);
		pluginMap.put(ConstantsPlugin.INSTANTMESSENGER_FILESTORE_PATH, INSTANTMESSENGER_FILESTORE_PATH);
		pluginMap.put(BBP_APIACCOUNT_APIID_PARAM, BBP_APIACCOUNT_APIID);
		pluginMap.put(BBP_APIACCOUNT_APIKEY_PARAM, BBP_APIACCOUNT_APIKEY);
		return pluginMap;
	}

}


src/de/citunius/businesslogic/calculator/ConstantsBot.java
/*
 * Copyright: (c) 2015-2019, Citunius GmbH. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * Licence: This program contains proprietary and trade secret information of Citunius GmbH.
 *          Copyright notice is precautionary only and does not evidence any actual or intended 
 *          publication of such program
 *          See: https://www.citunius.de/en/legal
 *
 * Requires: JDK 1.8+
 * $Id: ConstantsCalculator.java 12 2020-01-21 11:40:54Z  $
 *
 */
package de.citunius.businesslogic.calculator;

/**
 * This class holds all constants
 * 
 * @author me
 * @version %I%, %G%
 * @since   1.0
 */

public class ConstantsBot {
	// Plugin name
	public static final String PLUGIN_NAME = "GM.Calculator";
	
	// Event Types and Event Names
	public static final String EVENTTYPE_MSGGATEWAY_NOTIFICATION = "Notification";
	public static final String EVENTNAME_MSGGATEWAY_NOTIFICATION_PLAINTEXT = "plainText";
	
	// View names to know which telegram message template should be used for an incoming system event
	public static final String VIEW_DEFAULT = "default";
}


src/de/citunius/businesslogic/calculator/botbuilder/BotBuilder.java
/*
 * Copyright: (c) 2015-2019, Citunius GmbH. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * Licence: This program contains proprietary and trade secret information of Citunius GmbH.
 *          Copyright notice is precautionary only and does not evidence any actual or intended 
 *          publication of such program
 *          See: https://www.citunius.de/en/legal
 *
 * Requires: JDK 1.8+
 * $Id: BotBuilder.java 12 2020-01-21 11:40:54Z  $
 *
 */
package de.citunius.businesslogic.calculator.botbuilder;

import java.lang.invoke.MethodHandles;
import java.util.HashMap;

import org.apache.log4j.Logger;

import de.citunius.businesslogic.calculator.services.CalculatorService;
import de.citunius.businesslogicapi.common.ConstantsPlugin;
import de.citunius.businesslogicapi.common.botbuilder.BotBuilderService;
import de.citunius.businesslogicapi.common.services.LocalisationService;

/**
 * BotBuilderService is the class represents the services for the Bot Builder Model.
 * This service calls the use case specific functions of this business bot plugin.
 * This class is called using Java reflections.
 *
 * @author me
 * @version %I%, %G%
 * @since   1.0
*/
public class BotBuilder implements BotBuilderService {
	static Logger logger = Logger.getLogger(MethodHandles.lookup().lookupClass());
	
	// Defines the syntax for function parameter field
	public static final String REQUIRED_PLUGINMAP_FUNCTION_PARAMS_FIELD = "REQUIRED_PLUGINMAP_FUNCTION_PARAMS_";
	
	/**
	 * Class constructor
	 * 
	 * @param pluginMap
	 * @param ls
	 * @param defaultLanguage
	 */
	public BotBuilder() {
		
	}
	
	/**
	 * Bot builder function for the dialogue designer
	 * 
	 * <p>This function calls the calculator service in order to 
	 * calculate the math task. The result is formatted as returned 
	 * to the business bot platform for further processing</p>
	 * 
	 * @param pluginMap  the plugin hashmap
	 * @param ls  the language service instance
	 * @param defaultLanguage  the default user language
	 * @return  the result
	 */
	public String getMathResult(HashMap<String, String> pluginMap, LocalisationService ls, String defaultLanguage) {
		String result = null;
		
		if (pluginMap.containsKey(ConstantsPlugin.BOTBUILDER_FUNCTION_PARAM_1.toString())) {
			String mathTask = pluginMap.get(ConstantsPlugin.BOTBUILDER_FUNCTION_PARAM_1.toString());
			logger.info("mathTask: ["+mathTask+"]");
			
	    	CalculatorService cs = new CalculatorService(pluginMap);
    		if (cs.isMathTask(mathTask)) {
    			logger.debug("User message contains math"); 
    			double mathResult;
				try {
					mathResult = cs.calculate(mathTask);
					result = ls.getString("Calculator.Result", defaultLanguage)+": "+mathResult;
				} catch (Exception e) {
					logger.error(e);
				}    			
    		} else {
    			logger.info("User message does not contains math.");
    			result = ls.getString("Calculator.NoMath", defaultLanguage);
    		}
		} else {
			logger.warn("Math task not found in pluginMap -> expected as key ["+ConstantsPlugin.BOTBUILDER_FUNCTION_PARAM_1+"]");
		}
		
    	String replyMessage = "No reply content";
    	if (result != null) {
    		replyMessage = result;
    	} else {
    		replyMessage = ls.getString("Common.NoResultsFound", defaultLanguage);
    	}
    	return replyMessage;
	}
	
}


src/de/citunius/businesslogic/calculator/CalculatorBot.java
/*
 * Copyright: (c) 2015-2019, Citunius GmbH. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * Licence: This program contains proprietary and trade secret information of Citunius GmbH.
 *          Copyright notice is precautionary only and does not evidence any actual or intended 
 *          publication of such program
 *          See: https://www.citunius.de/en/legal
 *
 * Requires: JDK 1.8+
 * $Id: CalculatorBot.java 12 2020-01-21 11:40:54Z  $
 *
 */

package de.citunius.businesslogic.calculator;

import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;
import java.lang.invoke.MethodHandles;
import java.util.HashMap;
import java.util.Map;

import org.apache.log4j.Logger;
import org.json.JSONArray;
import org.json.JSONObject;

import de.citunius.bbp.im.gm.common.util.Emoji;
import de.citunius.bbp.im.gm.common.util.Util;
import de.citunius.bbp.im.gm.objects.Filter;
import de.citunius.bbp.im.gm.objects.Message;
import de.citunius.bbp.im.gm.objects.PluginReturnMessage;
import de.citunius.bbp.im.gm.objects.WebMessage;
import de.citunius.bbp.objects.AnonymousUserAccount;
import de.citunius.bbp.objects.ChatbotCallback;
import de.citunius.bbp.objects.MobileUserAccount;
import de.citunius.bbp.objects.Userstate;
import de.citunius.businesslogic.calculator.services.CalculatorService;
import de.citunius.businesslogicapi.common.ConstantsCommon;
import de.citunius.businesslogicapi.common.ConstantsPlugin;
import de.citunius.businesslogicapi.common.Utilities;
import de.citunius.businesslogicapi.common.services.LocalisationService;
import de.citunius.businesslogicapi.common.services.TemplateService;
import de.citunius.businesslogicapi.common.services.WebService;
import freemarker.template.Template;
import freemarker.template.TemplateException;

/**
 * CalculatorBotHandlers is the class represents the service bot calculator use case.
 * This service bot interacts with the mobile user and manages all user requests and responses.
 * 
 * Moreover, this service bot remembers the current user state (user menu),
 * composes message views and menus for the mobile user.
 *
 * @author me
 * @version %I%, %G%
 * @since   1.0
*/
public class CalculatorBot {
	static Logger logger = Logger.getLogger(MethodHandles.lookup().lookupClass());
	private static String defaultLanguage = "en";
    public static LocalisationService ls = null;
    public String tenantId = null;
    public String accountId = null;
    public HashMap<String, String> pluginMap = null;
    
    public String apiId = null;
    public String apiKey = null;
    
    public CalculatorBot(String tenantId, String accountId, HashMap<String, String> pluginMap) {
        logger.info("called");
        this.tenantId = tenantId;
        this.accountId = accountId;
        this.pluginMap = pluginMap;
        
        this.apiId = Utilities.getAttributeFromPropertiesFile(pluginMap, ConstantsCommon.BUSINESSBOTPLATFORM_ACCOUNT_APIID);
        this.apiKey = Utilities.getAttributeFromPropertiesFile(pluginMap, ConstantsCommon.BUSINESSBOTPLATFORM_ACCOUNT_APIKEY);
        
        logger.info("tenantId: ["+tenantId+"] accountId: ["+accountId+"]");
        ls = new LocalisationService(pluginMap);
    }
    
    /**
     * 
     * @param message
     * @param chatId
     * @param filter
     * @param username
     * @return
     */
    public PluginReturnMessage managePageRequestForMessageObjectResult(Message message, String userId, Filter filter, String username) {    	
		try {
			// Create the corresponding telegram message
			if (null != filter.getView() && filter.getView().equals(ConstantsBot.VIEW_DEFAULT) ) {
				// Print results using the default template for telegram message
				return composePageForDefaultObjectList(userId, filter, message, username);
			} else {
				// Print results using standard telegram message
				return composePageForDefaultObjectList(userId, filter, message, username);
			}
		} catch (TemplateException e) {
			logger.error(e);
		} catch (IOException e) {
			logger.error(e);
		}
		return null;
    }
    
    /**
     * This function composes the page for the list of default objects using a template
     *
     * @param userId  the chat identifier
     * @param filter  the filter object
     * @param message  the message object
     * @return the <code>SendMessage</code> method
     * @exception IOException On input error.
     * @exception TemplateException On template error.
     * @see IOException
     * @see TemplateException
     */
    private PluginReturnMessage composePageForDefaultObjectList(String userId, Filter filter, Message message, String username) throws TemplateException, IOException {    	
    	TemplateService ts = new TemplateService(pluginMap);
	    Template templatePage = ts.getTemplate("list.ftl");
	    Map<String, Object> dataPage = new HashMap<String, Object>();
	    Writer outPage = new StringWriter();
	    dataPage.put("title", ls.getString("Generic.Results.PageTitle", defaultLanguage));
		dataPage.put("listoutput", getMessageObjectResultForDefaultList(userId, filter, message, username));
		templatePage.process(dataPage, outPage);
		
		PluginReturnMessage pluginReturnMessage = new PluginReturnMessage(null, outPage.toString(), null);
	    return pluginReturnMessage;	   
    }
    
    /**
     * This function composes the data for default objects
     *
     * @param userId  the chat identifier
     * @param job  the job object
     * @param message  the message object
     * @return the composed list of default objects
     */
    private String getMessageObjectResultForDefaultList(String userId, Filter filter, Message message, String username) {    	
    	String text = "";
	    logger.info("MessageResult field content: ["+message.getData().getResult()+"]");
	    
	    if (message.getData() != null && message.getData().getResult() != null) {
	    	JSONArray jsonArray = null;
	    	try {
				jsonArray = new JSONArray(message.getData().getResult());
			} catch(Exception e) {
				logger.error(e);
			}
			if (jsonArray != null && jsonArray.length() != 0) {
				logger.info("jsonArray: ["+jsonArray+"] jsonArray.length(): ["+jsonArray.length()+"]");
				for (int i = 0; i < jsonArray.length(); i++) {				
					WebMessage wm = new WebMessage(jsonArray.getJSONObject(i));
					logger.info("WebMessage: Owner: ["+wm.getOwner()+"] Content: ["+wm.getContent()+"]");
					text += String.format(ls.getString("NewMsgCompose.TextHeader", defaultLanguage), Util.replaceRestrictedChars((username != null ? username : "UnknownUser"))) + "\n\n";
					text += "--------------------------\n";
					text += Util.replaceRestrictedChars(wm.getContent())+"\n";		    	
			    	text += "--------------------------\n\n";
			    	text += ls.getString("NewMsgCompose.TextFooter", defaultLanguage);
				}
			} else {
				//text += ConstantsGeneric.ERROR_INTERNALPROCESSING;
				text += ls.getString("Common.NoResultsFound", defaultLanguage);
			}
	    } else {
	    	text += ls.getString("Common.NoResultsFound", defaultLanguage);
	    }
		// Add attachments
		if (message.getAudio() != null) {
			logger.info("Adding audio to message");
			text += ConstantsCommon.FILEUUID_TAG_BEGIN + message.getAudio().getFileUUID() + ConstantsCommon.FILEUUID_TAG_END;
		}
		if (message.getDocument() != null) {
			logger.info("Adding document to message");
			text += ConstantsCommon.FILEUUID_TAG_BEGIN + message.getDocument().getFileUUID() + ConstantsCommon.FILEUUID_TAG_END;
		}
		if (message.getLocation() != null) {
			logger.info("Adding location to message");
			text += ConstantsCommon.LOCATION_TAG_BEGIN + message.getLocation().getLatitude() + ConstantsCommon.LOCATION_TAG_DELIMITER + message.getLocation().getLongitude() + ConstantsCommon.LOCATION_TAG_END;
		}
		if (message.getPhoto() != null && message.getPhoto().size() != 0) {
			logger.info("Adding photos to message");
			for (int p=0; p<message.getPhoto().size(); p++) {
				text += ConstantsCommon.FILEUUID_TAG_BEGIN + message.getPhoto().get(p).getFileUUID() + ConstantsCommon.FILEUUID_TAG_END;
			}			
		}
		if (message.getVideo() != null) {
			logger.info("Adding video to message");
			text += ConstantsCommon.FILEUUID_TAG_BEGIN + message.getVideo().getFileUUID() + ConstantsCommon.FILEUUID_TAG_END;
		}
		if (message.getVoice() != null) {
			logger.info("Adding voice to message");
			text += ConstantsCommon.FILEUUID_TAG_BEGIN + message.getVoice().getFileUUID() + ConstantsCommon.FILEUUID_TAG_END;
		}
		
		return text;		
    }

	private static String getCommandAbout() {
	    return String.format(ls.getString("Common.About", defaultLanguage), Emoji.ABOUT.toString());
	}

	private static String getCommandBack() {
	    return String.format(ls.getString("Common.Back", defaultLanguage), Emoji.BACK_WITH_LEFTWARDS_ARROW_ABOVE.toString());
	}

	private static String getCommandCancel() {
	    return String.format(ls.getString("Common.Cancel", defaultLanguage), Emoji.DISCUSSION_NOREPLY.toString());
	}

	private static String getCommandLanguages() {
	    return String.format(ls.getString("Languages", defaultLanguage), Emoji.GLOBE_WITH_MERIDIANS.toString());
	}

	private String getMessageLanguage() {
	    String baseString = ls.getString("Language.SelectLanguage", defaultLanguage);
	    return String.format(baseString, defaultLanguage);
	}

	/**
	 * 
	 * @param tenantId
	 * @param accountId
	 * @param jsonMessage
	 * @param jsonMobileUserAccount
	 * @return
	 */
	public String handleIncomingMessage(String tenantId, String accountId, String jsonMessage, boolean anonymousUserAccountExists, String jsonAnonymousUserAccount, boolean mobileUserAccountExists, String jsonMobileUserAccount) {		
		logger.info("Plugin: New jsonMessage: ["+jsonMessage+"]");
		JSONObject jsonObjectMessage = new JSONObject(jsonMessage);
		Message message = new Message(jsonObjectMessage);
		MobileUserAccount mobileUserAccount = null;
		if (mobileUserAccountExists) {
			JSONObject jsonObjectMobileUserAccount = new JSONObject(jsonMobileUserAccount);
			mobileUserAccount = new MobileUserAccount(jsonObjectMobileUserAccount);
		}
		logger.info("Plugin: New message: ["+message.getText()+"]");
		logger.info("From UserId: ["+message.getFrom().getId()+"] First-/Lastname: ["+message.getFrom().getFirstName()+" "+message.getFrom().getLastName()+"]");
		logger.info("message text: ["+message.hasText()+"]");
		
		logger.info("Message received from user: ["+message.getFrom().getUserName()+"]");
		if (mobileUserAccount != null) {
			logger.info("Username: ["+mobileUserAccount.getUsername()+"]");
		}
		
		String replySubject = ls.getString("Common.DefaultReplySubject", defaultLanguage);
    	String replyMessage = "No reply content";
    	String replyErrorMessage = ls.getString("Common.Error", defaultLanguage);  	
		
    	try {
			Userstate userstate = Utilities.getUserState(tenantId, accountId, ConstantsPlugin.PLUGIN_NAME, pluginMap, message.getFrom().getUserName());
			// Get user language to default language
	    	if (userstate != null && userstate.getUserLanguageCode() != null) {
	    		defaultLanguage = userstate.getUserLanguageCode();
	    		logger.info("Set user language code to ["+userstate.getUserLanguageCode()+"]");
	    	} else {
	    		logger.info("No user state found for user ["+message.getFrom().getUserName()+"]@["+ConstantsPlugin.PLUGIN_NAME+"]");
	    	}
		} catch(Exception e) {
			logger.error("Exception: "+e.getMessage());
			logger.error(e);
			return new PluginReturnMessage(replySubject, replyErrorMessage, null).toJson().toString();
		}		

    	if (message.getText() != null && message.getText().length() != 0) {
    		CalculatorService cs = new CalculatorService(pluginMap);
    		if (cs.isMathTask(message.getText())) {
    			logger.debug("User message contains math"); 
    			double result;
				try {
					result = cs.calculate(message.getText());
					replyMessage = ls.getString("Calculator.Result", defaultLanguage)+": "+result;
				} catch (Exception e) {
					logger.error(e);
				}    			
    		} else {
    			logger.info("User message does not contains math.");
    			replyMessage = ls.getString("Calculator.NoMath", defaultLanguage);
    		}
    	} else {
    		logger.info("User message does not contains math.");
    		replyMessage = ls.getString("Calculator.NoMath", defaultLanguage);
    	}    	
    	return new PluginReturnMessage(replySubject, replyMessage, null).toJson().toString();
	}
	
	/**
	 * 
	 * @param message
	 * @param filter
	 * @param mobileUserAccount
	 * @return
	 */
	public String handleIncomingCallback(String tenantId, String accountId, String message, boolean anonymousUserAccountExists, String jsonAnonymousUserAccount, boolean mobileUserAccountExists, String jsonMobileUserAccount) {
		logger.info("handleIncomingCallback() called");
		
		String replyMessage = "NOT_IMPLEMENTED";
		
		PluginReturnMessage pluginReturnMessage = new PluginReturnMessage(null, "Callback message: ["+replyMessage+"]", null);
		return pluginReturnMessage.toJson().toString();
	}
	
	
	/**
	 * Call webservice to schedule a callback for this plugin
	 * 
	 * @param mobileUserAccount  the mobile user account object
	 * @return
	 */
	public boolean scheduleCallback(boolean anonymousUserAccountExists, AnonymousUserAccount anonymousUserAccount, boolean mobileUserAccountExists, MobileUserAccount mobileUserAccount) {
		String cronSchedule = "0 0/1 * * * ? *";
		String callbackMessage = "reminder-xy";
		ChatbotCallback chatbotCallback = new ChatbotCallback(
				Integer.parseInt(pluginMap.get(ConstantsPlugin.BOTID)), 
				Integer.parseInt(pluginMap.get(ConstantsPlugin.PLUGIN_ID)), 
				anonymousUserAccountExists,
				anonymousUserAccount,
				mobileUserAccountExists,
				mobileUserAccount, 
				cronSchedule, 
				callbackMessage);
		
		// Register chatbot callback to BBP webservices
		String receiptId = null;
		try {
			WebService ws = new WebService(pluginMap);
			receiptId = ws.registerChatbotCallback(chatbotCallback);
		} catch (Exception e) {
			logger.error("Unable to register chatbot callback ["+chatbotCallback.getCallbackMessage()+"] to BBP webservices: "+e.getMessage());
			return false;
		}
		if (receiptId != null && receiptId.length() != 0) {
			logger.info("Chatbot callback ["+chatbotCallback.getCallbackMessage()+"] has been registed to BBP webservices successfully. Receipt Id: ["+receiptId+"]");
		} else {
			logger.error("Failed to register chatbot callback ["+chatbotCallback.getCallbackMessage()+"] to BBP webservices");
			return false;
		}
		return true;
	}

	/**
	 * 
	 * @param message
	 * @param filter
	 * @param mobileUserAccount
	 * @return
	 */
	public String sendTextMessage(String tenantId, String accountId, String jsonMessage, String jsonFilter, boolean anonymousUserAccountExists, String jsonAnonymousUserAccount, boolean mobileUserAccountExists, String jsonMobileUserAccount) {
		logger.info("sendTextMessage() called");
		JSONObject jsonObjectMessage = new JSONObject(jsonMessage);
		Message message = new Message(jsonObjectMessage);
		JSONObject jsonObjectFilter = new JSONObject(jsonFilter);
		Filter filter = new Filter(jsonObjectFilter);
		
		logger.info("Plugin: New message: ["+message.getText()+"] from UserId: ["+message.getFrom().getUserName()+"] First-/Lastname: ["+message.getFrom().getFirstName()+" "+message.getFrom().getLastName()+"]");
		AnonymousUserAccount anonymousUserAccount = null;
		MobileUserAccount mobileUserAccount = null;
		String userId = null;
		if (anonymousUserAccountExists) {
			JSONObject jsonObjectAnonymousUserAccount = new JSONObject(jsonAnonymousUserAccount);
			anonymousUserAccount = new AnonymousUserAccount(jsonObjectAnonymousUserAccount);
			logger.info("Managing request with AnonymousUserAccount: userId: ["+anonymousUserAccount.getUserId()+"]");
			userId = anonymousUserAccount.getUserId();
		}
		if (mobileUserAccountExists) {
			JSONObject jsonObjectMobileUserAccount = new JSONObject(jsonMobileUserAccount);
			mobileUserAccount = new MobileUserAccount(jsonObjectMobileUserAccount);
			logger.info("Managing request with MobileUserAccount: Id: ["+mobileUserAccount.getId()+"]");
			userId = Utilities.getMobileAppSettingValue(Utilities.getMobileAppSettingParameterNameUserId(tenantId, accountId, pluginMap), mobileUserAccount.getPrimaryMobileDevice().getPrimaryMobileApp().getMobileAppSettings());
		}
		
		if (userId == null) {
			logger.info("userId not found");
			return null;
		} else {
			logger.info("Using userId: ["+userId+"]");
		}
		
		PluginReturnMessage pluginReturnMessage = managePageRequestForMessageObjectResult(message, userId, filter, (mobileUserAccountExists ? mobileUserAccount.getUsername() : null));
		// Manage attachments
		PluginReturnMessage pluginReturnMessageWithAttachments = manageAttachments(pluginReturnMessage, message);		
		return pluginReturnMessageWithAttachments.toJson().toString();
    }
	
	/**
	 * Manage attachments 
	 * 
	 * @param pluginReturnMessage  the plugin return message
	 * @return the plugin return message with attachments
	 */
	public PluginReturnMessage manageAttachments(PluginReturnMessage pluginReturnMessage, Message message) {
		// Add media such as photos and documents (from incoming message)
		if (message.getAudio() != null) {
			logger.info("Adding audio to pluginReturnMessage");
			pluginReturnMessage.setAudio(message.getAudio());
		}
		if (message.getDocument() != null) {
			logger.info("Adding document to pluginReturnMessage");
			pluginReturnMessage.setDocument(message.getDocument());
		}
		if (message.getLocation() != null) {
			logger.info("Adding location to pluginReturnMessage");
			pluginReturnMessage.setLocation(message.getLocation());
		}
		if (message.getPhoto() != null) {
			logger.info("Adding photo to pluginReturnMessage");
			pluginReturnMessage.setPhoto(message.getPhoto());
		}
		if (message.getVideo() != null) {
			logger.info("Adding video to pluginReturnMessage");
			pluginReturnMessage.setVideo(message.getVideo());
		}
		if (message.getVoice() != null) {
			logger.info("Adding voice to pluginReturnMessage");
			pluginReturnMessage.setVoice(message.getVoice());
		}		
		return pluginReturnMessage;
	}
}
